README.md
<!-- markdownlint-disable no-inline-html -->
<!-- markdownlint-disable first-line-heading -->
<div align="center">
  <a href="https://github.com/lbragile/ColorMaster">
    <img src="public/logo.png" width="200" height="200" alt="ColorMaster Logo" />
  </a>
</div>

<div align="center">
  <a href="https://www.npmjs.com/package/colormaster">
    <img alt="package version" src="https://img.shields.io/npm/v/colormaster?label=NPM&logo=npm&style=flat-square" />
  </a>
  <a href="https://www.npmjs.com/package/colormaster">
    <img alt="minified & gzipped size" src="https://img.shields.io/bundlephobia/minzip/colormaster?label=Min%20%26%20Gzip&logo=npm&style=flat-square" />
  </a>
   <a href="https://www.npmjs.com/package/colormaster">
    <img alt="tree shaking" src="https://img.shields.io/static/v1?label=Tree%20Shaking&message=yes&color=%23007EC6&logo=npm&style=flat-square" />
  </a>
  <a href="https://github.com/lbragile/ColorMaster/actions">
    <img alt="build status" src="https://img.shields.io/github/workflow/status/lbragile/ColorMaster/Testing?color=%2300D800&label=Build&logo=github&style=flat-square" />
  </a>
  <a href="https://codecov.io/gh/lbragile/ColorMaster">
    <img alt='code coverage' src="https://img.shields.io/codecov/c/github/lbragile/colormaster/develop?color=%2300D800&label=Coverage&logo=codecov&style=flat-square"/>
  </a>
  <a href="https://www.codefactor.io/repository/github/lbragile/colormaster">
    <img alt="code quality" src="https://img.shields.io/codefactor/grade/github/lbragile/ColorMaster/develop?color=%2300D800&label=Code%20Quality&logo=codefactor&style=flat-square" />
  </a>
</div>

<div align="center">
  <strong>ColorMaster</strong> is a powerful and fully typed module for all your coloring needs.
</div>
<!-- markdownlint-enable first-line-heading -->
<!-- markdownlint-enable no-inline-html -->

## 💎 Features

- 📦 **Compact**: Just ~2.5 KB [minified & gzipped](https://bundlephobia.com/package/colormaster@latest)
- 👌 **Straightforward**: Clear & intuitive function names with chainable API
- 🔫 **Bulletproof**: Entirely TypeScript (strict mode) and very thoroughly tested
- ✍ **Strongly Typed**: Includes [types](#-strongly-typed) out of the box
- 🏗 **Extendable**: [Plugins](#-general-plugins) allow you to extend upon core functionality as needed
- 🌳 **Tree Shakable**: Only adds functions that you use to reduce your project's size
- 📑 **CSS Compliant**: Adheres to [CSS color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) specifications
- 🚂 **Works Everywhere**: Browser (client) and NodeJS (server) support
- 🐣 **Dependency Free**: Full control of features & enhancements

&nbsp;

## ⭐ Getting Started

```markdown
npm install colormaster
```

Then simply start using **ColorMaster** in your project:

<!-- markdownlint-disable no-inline-html -->
<details open>
<summary><b>RGBA</b> Color Space</summary>

```ts
// note that the import name can be anything (default export), we use "CM" for brevity
import CM from "colormaster";

// object argument
CM({ r: 128, g: 128, b: 128, a: 0.5 }).stringRGB(); // "rgba(128, 128, 128, 0.5)"
CM({ r: 128, g: 128, b: 128, a: 0.5 }).stringRGB({ alpha: false }); // "rgb(128, 128, 128)"
CM({ r: 128, g: 128, b: 128, a: 0.5 }).alphaTo(0.8).stringRGB(); // "rgba(128, 128, 128, 0.8)"

// string argument
CM("rgba(128, 128, 128, 0.5)").stringRGB(); // "rgba(128, 128, 128, 0.5)"
CM("rgba(128, 128, 128, 0.5)").stringRGB({ alpha: false }); // "rgb(128, 128, 128)"
CM("rgba(128, 128, 128, 0.5)").alphaTo(0.8).stringRGB(); // "rgba(128, 128, 128, 0.8)"
CM("rgba(128, 128, 128, 0.5)").stringRGB(); // "rgba(128, 128, 128, 0.5)"
CM("rgba(50%, 50%, 50%, 50%)").stringRGB(); // "rgba(128, 128, 128, 0.5)"
CM("rgba(50% 50% 50% 50%)").stringRGB(); // "rgba(128, 128, 128, 0.5)"
CM("rgba(50% 50% 50% / 50%)").stringRGB(); // "rgba(128, 128, 128, 0.5)"
CM("rgba(50% 50% 50% / 0.5)").stringRGB(); // "rgba(128, 128, 128, 0.5)"
```

</details>

<details>
<summary><b>HSLA</b> Color Space</summary>

```ts
import CM from "colormaster";

// object argument
CM({ h: 300, s: 50, l: 60, a: 0.5 }).stringHSL(); // "hsla(300, 50%, 60%, 0.5)"
CM({ h: 300, s: 50, l: 60, a: 0.5 }).stringHSL({ alpha: false }); // "hsl(300, 50%, 60%)"
CM({ h: 300, s: 50, l: 60, a: 0.5 }).alphaTo(0.8).stringHSL(); // "hsla(300, 50%, 60%, 0.8)"

// string argument
CM("hsla(300, 50%, 60%, 0.5)").stringHSL(); // "hsla(300, 50%, 60%, 0.5)"
CM("hsla(300, 50%, 60%, 0.5)").stringHSL({ alpha: false }); // "hsl(300, 50%, 60%)"
CM("hsla(300, 50%, 60%, 0.5)").alphaTo(0.8).stringHSL(); // "hsla(300, 50%, 60%, 0.8)"
CM("hsla(50%, 50%, 60%, 50%)").stringHSL(); // "hsla(180, 50%, 60%, 0.5)"
CM("hsla(50% 50% 60% 50%)").stringHSL(); // "hsla(180, 50%, 60%, 0.5)"
CM("hsla(50% 50% 60% / 50%)").stringHSL(); // "hsla(180, 50%, 60%, 0.5)"
CM("hsla(50% 50% 60% / 0.5)").stringHSL(); // "hsla(180, 50%, 60%, 0.5)"
```

</details>

<details>
<summary><b>HEXA</b> Color Space</summary>

```ts
import CM from "colormaster";

// object argument
CM({ r: "45", g: "67", b: "89", a: "AB" }).stringHEX(); // "#456789AB"
CM({ r: "45", g: "67", b: "89", a: "AB" }).stringHEX({ alpha: false }); // "#456789"
CM({ r: "45", g: "67", b: "89", a: "AB" }).alphaTo("CC").stringHEX(); // "#456789CC"

// string argument
CM("#456789AB").stringHEX(); // "#456789AB" ← 8-Digit Hex Input
CM("#456789").stringHEX(); // "#456789FF" ← 6-Digit Hex Input
CM("#4567").stringHEX(); // "#44556677" ← 4-Digit Hex Input
CM("#456").stringHEX(); // "#445566FF" ← 3-Digit Hex Input
CM("#456789AB").stringHEX({ alpha: false }); // "#456789"
CM("#456789AB").alphaTo("CC").stringHEX(); // "#456789CC"
```

**Note:** HEXA string are always returned in _upperCase_ by **ColorMaster**. If you prefer _lowerCase_ strings,
you can simply use (chain) the built in `toLowerCase()`. More information is available [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLowerCase)

</details>
<br/>

<!-- markdownlint-enable no-inline-html -->

**Note:** Many more color spaces are provided via [plugins](#-color-space-plugins) to make **ColorMaster** compact.

&nbsp;

## 🎯 Playground

**ColorMaster** also has a [playground](https://colormaster.vercel.app) where key functionality is illustrated.
Those that are interested can play around with the elegant user interface for daily color related needs,
grab the corresponding ColorMaster snippet from each page's `code` button, or even use our [REST API](https://github.com/lbragile/ColorMasterDemo/blob/develop/rest.http) for quick results to common requests.

&nbsp;

## 🔥 API

### Object Instantiation (Color Generation)

<!-- markdownlint-disable no-inline-html -->

<details>
<summary><code>ColorMaster(color)</code></summary>

Parses a given input color so that **ColorMaster** can perform necessary operations to manipulate/transform that input

```ts
import CM from "colormaster";

CM("rgba(255, 0, 0, 1)").stringRGB(); // "rgba(255, 0, 0, 1.0)"
CM({ r: 255, g: 0, b: 0, a: 1 }).stringRGB(); // "rgba(255, 0, 0, 1.0)"

CM("hsla(120, 100%, 50%, 1)").stringHSL(); // "hsla(120, 100%, 50%, 1.0)"
CM({ h: 120, s: 100, l: 50, a: 1 }).stringHSL(); // "hsla(120, 100%, 50%, 1.0)"

// Supports 3-digit, 4-digit, 6-digit, and 8-digit HEX(A) input.
CM("#456789").stringHEX(); // "#456789FF"
CM({ r: "45", g: "67", b: "89", a: "AB" }).stringHEX(); // "#456789AB"

// requires name plugin
import CM, { extendPlugins } from "colormaster";
import NamePlugin from "colormaster/plugins/name";

extendPlugins([NamePlugin]);

CM("red").stringHEX(); // "#FF0000FF"
```

</details>

<details>
<summary><code>random()</code></summary>

Generates a random RGBA color.

```ts
import { random } from "colormaster";

random().stringRGB(); // "rgba(255, 0, 0, 1.0)"
random()().stringHSL(); // "hsla(0, 100%, 50%, 1.0)"
random().stringHEX(); // "#FF0000FF"
```

</details>

<!-- markdownlint-enable no-inline-html -->

&nbsp;

### Formatting

<!-- markdownlint-disable no-inline-html -->

<details>
<summary><code>stringRGB(opts?)</code></summary>

Generates a human readable RGBA color space string from a given color instance

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").stringRGB(); // "rgba(255, 0, 0, 1)"
CM("rgba(255, 0, 0, 1)").stringRGB({ alpha: false }); // "rgba(255, 0, 0)"
CM("rgba(200.1234, 100.1234, 50.1234, 0.61234)").stringRGB({ precision: [1, 2, 3, 4] }); // "rgba(200.1, 100.12, 50.123, 0.6123)"
```

</details>

<details>
<summary><code>stringHEX(opts?)</code></summary>

Generates a human readable HEXA color space string from a given color instance

Note: this method always returns an 8-digit hex string (you do not need to/cannot specify precision)

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").stringHEX(); // "#FF0000FF"
CM("rgba(255, 0, 0, 1)").stringHEX({ alpha: false }); // "#FF0000"
```

</details>

<details>
<summary><code>stringHSL(opts?)</code></summary>

Generates a human readable HSLA color space string from a given color instance

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").stringHSL(); // "hsl(0, 100%, 50%, 1)"
CM("rgba(255, 0, 0, 1)").stringHSL({ alpha: false }); // "hsl(0, 100%, 50%)"
CM("hsla(200.1234, 75.1234, 50.1234, 0.61234)").stringHSL({ precision: [1, 2, 3, 4] }); // "hsla(200.1, 100.12%, 50.123%, 0.6123)"
```

</details>

<details>
<summary><code>name()</code> (<b>name</b> plugin)</summary>

Gets the color table <a href='https://www.rapidtables.com/web/color/RGB_Color.html'>CSS/HTML name</a> for a given color

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").name(); // "red"
CM("rgba(0, 255, 0, 1)").name(); // "green"
CM("rgba(0, 0, 255, 1)").name(); // "blue"
```

</details>
<br/>
<!-- markdownlint-enable no-inline-html -->

**Note:** More color space string formation methods (`string<COLOR-SPACE>`) are available with each [Color Space Plugin](#-color-space-plugins).

&nbsp;

### Channel/Instance Level Information

<!-- markdownlint-disable no-inline-html -->

<details>
<summary><code>format</code></summary>

Retrieve the format of color instance (what is the color space)

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").format; // "rgb"
```

</details>

<details>
<summary><code>red</code></summary>

Get the "red" channel value for the color instance.

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").red; // 255
CM("hsla(0, 100%, 50%, 1)").red; // 255
CM("#F00F").red; // "FF"
```

</details>

<details>
<summary><code>green</code></summary>

Get the "green" channel value for the color instance.

```ts
import CM from "colormaster";
CM("rgba(0, 255, 0, 1)").green; // 255
CM("hsla(120, 100%, 50%, 1)").green; // 255
CM("#0F0F").green; // "FF"
```

</details>

<details>
<summary><code>blue</code></summary>

Get the "blue" channel value for the color instance.

```ts
import CM from "colormaster";
CM("rgba(0, 0, 255, 1)").blue; // 255
CM("hsla(240, 100%, 50%, 1)").blue; // 255
CM("#00FF").blue; // "FF"
```

</details>

<details>
<summary><code>alpha</code></summary>

Get the "alpha" channel value for the color instance.

```ts
import CM from "colormaster";
CM("rgba(0, 0, 255, 1)").alpha; // 1
CM("hsla(240, 100%, 50%, 1)").alpha; // 1
CM("#FF0F").alpha; // "FF"
```

</details>

<details>
<summary><code>hue</code></summary>

Get the "hue" channel value for the color instance.

```ts
import CM from "colormaster";
CM("rgba(0, 0, 255, 1)").hue; // 240
CM("hsla(240, 100%, 50%, 1)").hue; // 240
CM("#00F").hue; // 240
```

</details>

<details>
<summary><code>saturation</code></summary>

Get the "saturation" channel value for the color instance.

```ts
import CM from "colormaster";
CM("rgba(0, 0, 255, 1)").saturation; // 100
CM("hsla(240, 100%, 50%, 1)").saturation; // 100
CM("#00F").saturation; // 100
```

</details>

<details>
<summary><code>lightness</code></summary>

Get the "lightness" channel value for the color instance.

```ts
import CM from "colormaster";
CM("rgba(0, 0, 255, 1)").lightness; // 50
CM("hsla(240, 100%, 50%, 1)").lightness; // 50
CM("#00F").lightness; // 50
```

</details>

<!-- markdownlint-enable no-inline-html -->

&nbsp;

### Objects From Color Instance

<!-- markdownlint-disable no-inline-html -->

<details>
<summary><code>rgba()</code></summary>

Retrieve the object corresponding to the color instance

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").rgba(); // "{ r: 255, g: 0, b: 0, a: 1 }"
```

</details>

<details>
<summary><code>hsla()</code></summary>

Retrieve the object corresponding to the color instance

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").hsla(); // "{ h: 0, s: 100, l: 50, a: 1 }"
```

</details>

<details>
<summary><code>hexa(opts?)</code></summary>

Retrieve the object corresponding to the color instance

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").hexa(); // "{ r: "FF", g: "00", b: "00", a: "FF" }"
```

</details>
<br/>
<!-- markdownlint-enable no-inline-html -->

**Note:** More color space object formation methods (`<color-space>a`) are available with each [Color Space Plugin](#-color-space-plugins).

&nbsp;

### Manipulation

<!-- markdownlint-disable no-inline-html -->

<details>
<summary><code>hueTo(value | cssName)</code></summary>

Changes the "hue" channel value TO a given input value (done in HSLA space and converted back to RGBA space)

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").hueTo(120).stringRGB(); // "rgba(0, 255, 0, 1.0)"
CM("hsla(0, 100%, 50%, 1)").hueTo(120).stringHSL(); // "hsla(120, 100%, 50%, 1.0)"
CM("#f00").hueTo("green").stringHEX(); // "#00FF00FF"
```

</details>

<details>
<summary><code>hueBy(delta)</code></summary>

Changes the "hue" channel value BY a given delta (done in HSLA space and converted back to RGBA space)

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").hueBy(120).stringRGB(); // "rgba(0, 255, 0, 1.0)"
CM("hsla(0, 100%, 50%, 1)").hueBy(-30).stringHSL(); // "hsla(330, 100%, 50%, 1.0)"
CM("#f00").hueBy(240).stringHEX(); // "#0000FFFF"
```

</details>

<details>
<summary><code>alphaTo(value)</code></summary>

Changes the "alpha" channel value TO a given input value

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").alphaTo(0.5).stringRGB(); // "rgba(255, 0, 0, 0.5)"
CM("hsla(0, 100%, 50%, 1)").alphaTo(0.5).stringHSL(); // "hsla(0, 100%, 50%, 0.5)"
CM("#f00").alphaTo("AA").stringHEX(); // "#FF0000AA"
```

</details>

<details>
<summary><code>alphaBy(delta)</code></summary>

Changes the "alpha" channel value BY a given delta

```ts
import CM from "colormaster";
CM("rgba(255, 0, 0, 1)").alphaBy(-0.1).stringRGB(); // "rgba(255, 0, 0, 0.9)"
CM("hsla(0, 100%, 50%, 1)").alphaBy(-0.1).stringHSL(); // "hsla(0, 100%, 50%, 0.9)"
CM("#f009").alphaBy("23").stringHEX(); // "#FF0000BC"
```

</details>

<details>
<summary><code>saturationTo(value)</code></summary>

Sets the saturation channel of a color to <code>value</code>

```ts
import CM from "colormaster";
CM("hsla(0, 60%, 50%, 1)").saturationTo(75).stringHSL(); // "hsla(0, 75%, 50%, 1.0)"
```

</details>

<details>
<summary><code>saturateBy(delta)</code></summary>

This adds pigmentation to a color, making it less gray.

```ts
import CM from "colormaster";
CM("hsla(0, 60%, 50%, 1)").saturateBy(20).stringHSL(); // "hsla(0, 80%, 50%, 1.0)"
```

</details>

<details>
<summary><code>desaturateBy(delta)</code></summary>

The opposite affect of <code>saturateBy(delta)</code>.
This removes pigmentation from a color making it more gray.

```ts
import CM from "colormaster";
CM("hsla(0, 60%, 50%, 1)").desaturateBy(20).stringHSL(); // "hsla(0, 40%, 50%, 1.0)"
```

</details>

<details>
<summary><code>lightnessTo(value)</code></summary>

Sets the lightness channel of a color to <code>value</code>

```ts
import CM from "colormaster";
CM("hsla(0, 60%, 50%, 1)").lightnessTo(75).stringHSL(); // "hsla(0, 60%, 75%, 1.0)"
```

</details>

<details>
<summary><code>lighterBy(delta)</code></summary>

This adds white to a color.

```ts
import CM from "colormaster";
CM("hsla(0, 60%, 50%, 1)").lighterBy(10).stringHSL(); // "hsla(0, 60%, 60%, 1.0)"
```

</details>

<details>
<summary><code>darkerBy(delta)</code></summary>

The opposite of <code>lighterBy(delta)</code>.
This adds black to a color.

```ts
import CM from "colormaster";
CM("hsla(0, 60%, 50%, 1)").darkerBy(10).stringHSL(); // "hsla(0, 60%, 40%, 1.0)"
```

</details>

<details>

CHANGELOG.md
# ColorMaster Change Log

All notable changes to the [ColorMaster](https://github.com/lbragile/ColorMaster) module will be documented in this file.

## v1.2.1 - September 06, 2021

### Fixed

- JSDoc of interfaces/class methods for better intellisense experience

---

## v1.2.0 - September 02, 2021

### Added

- `LUV`, `UVW`, `RYB` parsers & plugins with associated testing
- Added typeGuards for new color spaces & adjusted RGB/HEX type guards

### Fixed

- `RYBtoRGB` & `RGBtoRYB` conversions
- Parsing RegExp in all color spaces to be more robust with spaces before/after commas

### Changed

- Allow mixing in different color spaces

---

## v1.1.1 - August 24, 2021

### Added

- Links to JSDoc of parsers

### Fixed

- `stringCMYK()` & `stringXYZ()` and respective parsing
- Alpha channel value parsing

### Changed

- ColorMaster logo
- README color space plugins section and `stringRGB()`, `stringHSL()`, `stringHEX()`

---

## v1.1.0 - August 18, 2021

### Added

- Simplified logic of CM constructor
- `LAB[A]` `HSV[A]`, `HWB[A]`, `XYZ[A]`, `LCH[A]`, `CMYK[A]` parsers - along with respective plugins - and fully tested them
- Community documents such as [CONTRIBUTING.md](https://github.com/lbragile/ColorMaster/blob/master/.github/CONTRIBUTING.md)
- Issue & PR templates

### Fixed

- README to include more plugin information
- `typeGuard` logic to be more specific in scope
- Logic in parsers for object clamping

### Changed

- Webpack to not give sourcemaps
- Produce ES and CommonJS modules during bundling
- Test CI [GitHub Action](https://github.com/lbragile/ColorMaster/actions/workflows/test.yml) to use a matrix build - test common NodeJS version in parallel
- Exports field in `package.json` to show in [bundlephobia](https://bundlephobia.com/package/colormaster@latest)
- Adjusted Jest config to reduce CLI command

---

## v1.0.0 - August 15, 2021

### Added

- Ability to extend plugins
- [Tree Shaking](https://webpack.js.org/guides/tree-shaking/) with Webpack
- Bundle size analyzer (webpack plugin) ← only runs in production build
- Harmony plugin
- Mix in [LCHA](https://en.wikipedia.org/wiki/HCL_color_space) space (needs work to be more accurate)
- `isValid()` function & format getter is more robust
- Conversion helper functions to different color spaces
- Option to `hexa()` for rounding
- Tests for conversions
- Ability to extend parsers → allows to extend ColorMaster instantiation by CSS name
- LABA, LCHA, XYZA, HSVA conversion logic to and from RGBA
- `static Parsers` property to simplify logic
- Community related files & templates for issues/PRs
- Mixing (plugin) & tests
- Parsing tests
- Types to color Records so that they give intellisense

### Fixed

- Negative hue parsing
- Closest color logic by adding helper functions. Added `random()` outside of ColorMaster class. Start of `mix()` - working on RGBtoLAB & RGBtoLCH
- Improved parsers (RegExp)
- Fixed Jest config & adjusted coverage thresholds to be at **least 95%**
- README to reflect the refactor

### Changed

- Refactor to only use RGB behind the scene
- Parser functionality to be more general (better RegExp)
- All accessibility functions are now more general (independent of input format)
- `ColorMaster` class to make it default export a function that returns an instance
- Opts to be alpha for some method
- Separated parsers from conversion logic

### Removed

- Redundant typing on class method opts objects

---

## v0.1.0 - August 1, 2021

### Added

- Functionality and tests for:

  - `rotation`
  - `format`
  - `name`
  - `random`
  - `brightness`
  - `luminance`
  - `isDark`
  - `isLight`
  - `contrast`
  - `readableOn`
  - `equalTo`
  - `harmony` ([color harmonies](https://www.tigercolor.com/color-lab/color-theory/color-theory-intro.htm))
  - `isWarm`
  - `isCool`
  - `isTinted`
  - `isShaded`
  - `isToned`
  - `isPureHue`
  - `closestPureHue`
  - `closestWebSafe` ([web safe](https://www.rapidtables.com/web/color/Web_Safe.html))
  - `closestWarm`
  - `closestCool`

- README documentation for API
- Support for 3/4-digit HEX
- Size check GitHub action

### Fixed

- Tests to provide more coverage (100% coverage)
- `random()` and `fromName()` to generate RGB color instance which user can convert using `hsl()` or `hex()`
- Edge cases in `colormaster` class when generating color instances
- GitHub actions scripts

### Changed

- Getter & Setter function names to be more general
- `name()` function to search by name exactly or approximately
- `HSLAFrom()` to allow user to use CSS Name instead of number for hue channel

### Removed

- Benchmark tests as they are not meaningful enough
- Packages related to benchmark script

---

## v0.0.7 - July 28, 2021

### Added

- HSLA and HEXA color-spaces (with tests)
- Implementation interfaces to reduce complexity of implementation scripts (less comments)
- Getter and Setter functions for arrays of object instances
- [Webpack](https://webpack.js.org/) to create a production build
- `npm dedupe` to the pre-commit hook and post-commit/pre-push hooks
- Functionality such as invert, saturate/desaturate, lightness/darkness, grayscale
- Full testing of saturation/de-saturation, lighter/darker, grayscale in all color-spaces
- NPM scripts for releasing a publication of the module - patch, minor, major

### Fixed

- Logic in HEXA & HSLA color-spaces that produced invalid conversions
- Logic in class methods to use getter and setter functions rather than accessing private variables

### Changed

- Type file extensions to `.d.ts` - adjusted npm scripts accordingly
- Release script to have pre and post
- Default options from `false` to `true` as returning alpha in string is more common
- Updated README & documentation

### Removed

- Throw statements (errors) by using simple validation/setting to default values

---

## v0.0.2 to v0.0.6 - July 24, 2021

### Added

- NPM scripts to handle the module publication in a smoother and more proper manner

---

## v0.0.1 - July 24, 2021

### Added


src/debug.ts
/**
 * This file is meant for quick debugging and is not included in the final build
 * Triggered by the following npm script: `npm run debug`
 */

import CM from "./index";

console.log("\n🔸 luminance\n");

console.log(CM.RGBAFrom("128, 128, 128, 1").luminance());
console.log(CM.RGBAFrom("128, 128, 128, 1").luminance({ percentage: true }));
console.log(CM.RGBAFrom("204, 221, 238, 1").luminance({ percentage: true }));
console.log(CM.RGBAFrom("255, 255, 255, 1").luminance({ percentage: true }));

console.log("\n--------\n");

console.log("🔸 contrast\n");

console.log(CM.RGBAFrom("255, 255, 255, 1").contrast([255, 255, 0]));
console.log(CM.RGBAFrom("255, 0, 0, 1").contrast());
console.log(CM.RGBAFrom("255, 0, 0, 1").contrast([255, 255, 0], { ratio: true }));

console.log("\n--------\n");

console.log("🔸 readableOn\n");

console.log(CM.RGBAFrom("0, 0, 0, 1").readableOn());
console.log(CM.RGBAFrom("128, 128, 128, 1").readableOn());
console.log(CM.RGBAFrom("255, 255, 255, 1").readableOn([0, 0, 0]));
console.log(CM.RGBAFrom("230, 0, 0, 1").readableOn([255, 255, 71]));
console.log(CM.RGBAFrom("230, 0, 0, 1").readableOn([255, 255, 71], { ratio: "enhanced" }));
console.log(CM.RGBAFrom("230, 0, 0, 1").readableOn([255, 255, 71], { size: "large", ratio: "enhanced" }));

console.log("\n--------\n");

console.log("🔸 equalTo\n");

console.log(CM.RGBAFrom("0, 0, 0, 1").equalTo([0, 0, 0]));
console.log(CM.RGBAFrom("0, 0, 0, 1").equalTo([255, 255, 255]));
console.log(CM.RGBAFrom("0, 0, 0, 1").equalTo([0, 0, 0, 0.5]));
console.log(CM.RGBAFrom("230, 0, 0, 1").equalTo([230, 0, 0]));
console.log(CM.RGBAFrom("230, 0, 0, 1").equalTo(CM.RGBAFrom("230, 0, 0, 1")));
console.log(CM.RGBAFrom("230, 0, 0, 1").equalTo(CM.RGBAFrom("229, 0, 0, 1")));

console.log("\n--------\n");

console.log("🔸 random\n");

console.log(CM.random().string());
console.log(CM.random().string());
console.log(CM.random().string());
console.log(CM.random().string());
console.log(CM.random().string());
console.log(CM.random().string());
console.log(CM.random().string());
console.log(CM.random().string());

console.log("\n--------\n");

console.log("🔸 getter RGBA\n");

console.log(CM.RGBAFrom("230, 0, 0, 1").red);
console.log(CM.RGBAFrom("230, 0, 0, 1").blue);
console.log(CM.RGBAFrom("230, 0, 0, 1").green);
console.log(CM.RGBAFrom("230, 0, 0, 1").alpha);
console.log(CM.RGBAFrom("230, 0, 0, 1").hue);
console.log(CM.RGBAFrom("230, 0, 0, 1").saturation);
console.log(CM.RGBAFrom("230, 0, 0, 1").lightness);

console.log("\n🔸 getter HEXA\n");

console.log(CM.HEXAFrom("#aabbccdd").red);
console.log(CM.HEXAFrom("#aabbccdd").blue);
console.log(CM.HEXAFrom("#aabbccdd").green);
console.log(CM.HEXAFrom("#aabbccdd").alpha);
console.log(CM.HEXAFrom("#aabbccdd").hue);
console.log(CM.HEXAFrom("#aabbccdd").saturation);
console.log(CM.HEXAFrom("#aabbccdd").lightness);

console.log("\n🔸 getter HSLA\n");

console.log(CM.HSLAFrom("0, 50%, 50%").red);
console.log(CM.HSLAFrom("0, 50%, 50%").blue);
console.log(CM.HSLAFrom("0, 50%, 50%").green);
console.log(CM.HSLAFrom("0, 50%, 50%").alpha);
console.log(CM.HSLAFrom("0, 50%, 50%").hue);
console.log(CM.HSLAFrom("0, 50%, 50%").saturation);
console.log(CM.HSLAFrom("0, 50%, 50%").lightness);

console.log("\n--------\n");

console.log("🔸 name\n");

console.log(CM.RGBAFrom("0, 0, 0, 1").name());
console.log(CM.RGBAFrom("0, 0, 1, 1").name({ exact: false }));
console.log(CM.RGBAFrom("0, 0, 1, 1").name());
console.log(CM.HEXAFrom("#000000FF").name());
console.log(CM.HEXAFrom("#000001FF").name({ exact: false }));
console.log(CM.HEXAFrom("#000001FF").name());
console.log(CM.HSLAFrom("0, 0, 0, 1").name());
console.log(CM.HSLAFrom("0, 0, 1, 1").name({ exact: false }));
console.log(CM.HSLAFrom("0, 0, 1, 1").name());

console.log("\n--------\n");

console.log("🔸 name\n");
console.log("OG", CM.HSLAFrom("30, 50, 50, 1").string());

console.log("\nAnalogous\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("analogous")
    .map((color) => color.string())
);

console.log("\nComplementary\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("complementary")
    .map((color) => color.string())
);

console.log("\nSplit Complementary\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("split-complementary")
    .map((color) => color.string())
);

console.log("\nDouble Split Complementary\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("double-split-complementary")
    .map((color) => color.string())
);

console.log("\nTriad\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("triad")
    .map((color) => color.string())
);

console.log("\nRectangle\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("rectangle")
    .map((color) => color.string())
);

console.log("\nSquare\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("square")
    .map((color) => color.string())
);

console.log("\nMonochromatic - tints\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("monochromatic", { effect: "tints" })
    .map((color) => color.string())
);

console.log("\nMonochromatic - shades\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("monochromatic", { effect: "shades" })
    .map((color) => color.string())
);

console.log("\nMonochromatic - tones\n");

console.log(
  CM.HSLAFrom("30, 50, 50, 1")
    .harmony("monochromatic", { effect: "tones" })
    .map((color) => color.string())
);

console.log("\n--------\n");

console.log("Tinted/Shaded/Toned\n");

console.log(CM.HSLAFrom("100, 50, 50.1, 1").isTinted());
console.log(CM.HSLAFrom("100, 50, 50, 1").isTinted()); // false
console.log(CM.HSLAFrom("100, 50, 49.9, 1").isTinted()); // false
console.log(CM.HSLAFrom("100, 50, 50.1, 1").isShaded()); // false
console.log(CM.HSLAFrom("100, 50, 50, 1").isShaded()); // false
console.log(CM.HSLAFrom("100, 50, 49.9, 1").isShaded());
console.log(CM.HSLAFrom("100, 99.9, 50, 1").isToned());
console.log(CM.HSLAFrom("100, 100, 50, 1").isToned()); // false

console.log("\n--------\n");

console.log("Cool/Warm\n");

console.log(CM.HSLAFrom("254.9, 50, 50, 1").isCool());
console.log(CM.HSLAFrom("255, 50, 50, 1").isWarm());
console.log(CM.HSLAFrom("0, 50, 50, 1").isWarm());
console.log(CM.HSLAFrom("180, 50, 50, 1").isCool());
console.log(CM.HSLAFrom("74.9, 50, 50, 1").isWarm());
console.log(CM.HSLAFrom("75, 50, 50, 1").isCool());
console.log(CM.HSLAFrom("75, 50, 50, 1").isWarm()); // false

console.log("\n--------\n");

console.log("HSLA from CSS Name\n");

console.log(CM.HSLAFrom("red", 100, 50, 1).hue);
console.log(CM.HSLAFrom("orange", 100, 50, 1).hue);
console.log(CM.HSLAFrom("yellow", 100, 50, 1).hue);
console.log(CM.HSLAFrom("lime", 100, 50, 1).hue);
console.log(CM.HSLAFrom("green", 100, 50, 1).hue);
console.log(CM.HSLAFrom("green cyan", 100, 50, 1).hue);
console.log(CM.HSLAFrom("cyan", 100, 50, 1).hue);
console.log(CM.HSLAFrom("cyan blue", 100, 50, 1).hue);
console.log(CM.HSLAFrom("blue", 100, 50, 1).hue);
console.log(CM.HSLAFrom("purple", 100, 50, 1).hue);
console.log(CM.HSLAFrom("pink", 100, 50, 1).hue);
console.log(CM.HSLAFrom("pink red", 100, 50, 1).hue);
